Document middlware with minimal APIs (#27283)
dotnet/AspNetCore.Docs: aspnetcore/fundamentals/minimal-apis/middleware.md

---
title: Middleware with Minimal API applications

> author:
> description: Use middleware in Minimal API applications
> ms.author: 

Reviewer: ```suggestion
author: BrennanConroy
description: Use middleware in Minimal API applications
ms.author:  brecon
```

Accept my suggestion to fix build errors.

ms.date: 10/10/2022
monikerRange: '>= aspnetcore-7.0'
uid: fundamentals/minimal-apis/middleware
---
# Middleware in Minimal API apps

> `WebApplication` will automatically add the following middleware depending on certain conditions.

Reviewer: ```suggestion
`WebApplication` automatically adds the following middleware depending on certain conditions:
```

> * [`UseDeveloperExceptionPage`](/dotnet/api/microsoft.aspnetcore.diagnostics.developerexceptionpagemiddleware) will be added first if the [`HostingEnvironment`](xref:fundamentals/environments) is `"Development"`.

Reviewer: ```suggestion
* [`UseDeveloperExceptionPage`](/dotnet/api/microsoft.aspnetcore.diagnostics.developerexceptionpagemiddleware) is added first when the [`HostingEnvironment`](xref:fundamentals/environments) is `"Development"`.
```

> * [`UseRouting`](/dotnet/api/microsoft.aspnetcore.builder.endpointroutingapplicationbuilderextensions.userouting) will be added second if user code didn't already call `UseRouting` and if there are endpoints configured, for example `app.MapGet`.

Reviewer: ```suggestion
* [`UseRouting`](/dotnet/api/microsoft.aspnetcore.builder.endpointroutingapplicationbuilderextensions.userouting) is added second if user code didn't already call `UseRouting` and if there are endpoints configured, for example `app.MapGet`.
```

> * [`UseEndpoints`](/dotnet/api/microsoft.aspnetcore.builder.endpointroutingapplicationbuilderextensions.useendpoints) will be added at the end of the middleware pipeline if any endpoints are configured.

Reviewer: ```suggestion
* [`UseEndpoints`](/dotnet/api/microsoft.aspnetcore.builder.endpointroutingapplicationbuilderextensions.useendpoints) is added at the end of the middleware pipeline if any endpoints are configured.
```

> * [`UseAuthentication`](/dotnet/api/microsoft.aspnetcore.builder.authappbuilderextensions.useauthentication) will be added immediately after `UseRouting` if [`IAuthenticationSchemeProvider`](/dotnet/api/microsoft.aspnetcore.authentication.iauthenticationschemeprovider) can be detected in the service provider. `IAuthenticationSchemeProvider` is added by default when using [`AddAuthentication`](/dotnet/api/microsoft.extensions.dependencyinjection.authenticationservicecollectionextensions.addauthentication), and services are detected using [`IServiceProviderIsService`](/dotnet/api/microsoft.extensions.dependencyinjection.iserviceproviderisservice).

Reviewer: ```suggestion
* [`UseAuthentication`](/dotnet/api/microsoft.aspnetcore.builder.authappbuilderextensions.useauthentication) is added immediately after `UseRouting` if [`IAuthenticationSchemeProvider`](/dotnet/api/microsoft.aspnetcore.authentication.iauthenticationschemeprovider) can be detected in the service provider. `IAuthenticationSchemeProvider` is added by default when using [`AddAuthentication`](/dotnet/api/microsoft.extensions.dependencyinjection.authenticationservicecollectionextensions.addauthentication), and services are detected using [`IServiceProviderIsService`](/dotnet/api/microsoft.extensions.dependencyinjection.iserviceproviderisservice).
```

> * [`UseAuthorization`](/dotnet/api/microsoft.aspnetcore.builder.authorizationappbuilderextensions.useauthorization) will be added next if [`IAuthorizationHandlerProvider`](/dotnet/api/microsoft.aspnetcore.authorization.iauthorizationhandlerprovider) can be detected in the service provider. `IAuthorizationHandlerProvider` is added by default when using [`AddAuthorization`](/dotnet/api/microsoft.extensions.dependencyinjection.authenticationservicecollectionextensions.addauthentication), and services are detected using `IServiceProviderIsService`.

Reviewer: ```suggestion
* [`UseAuthorization`](/dotnet/api/microsoft.aspnetcore.builder.authorizationappbuilderextensions.useauthorization) is added next if [`IAuthorizationHandlerProvider`](/dotnet/api/microsoft.aspnetcore.authorization.iauthorizationhandlerprovider) can be detected in the service provider. `IAuthorizationHandlerProvider` is added by default when using [`AddAuthorization`](/dotnet/api/microsoft.extensions.dependencyinjection.authenticationservicecollectionextensions.addauthentication), and services are detected using `IServiceProviderIsService`.
```

> * Finally, user configured middleware and endpoints are added between `UseRouting` and `UseEndpoints`.

Reviewer: ```suggestion
* User configured middleware and endpoints are added between `UseRouting` and `UseEndpoints`.
```

> The following code is effectively what the automatic middleware being added to the application will look like.

Reviewer: ```suggestion
The following code is effectively what the automatic middleware being added to the application looks like.
```

```csharp
if (isDevelopment)
{
    app.UseDeveloperExceptionPage();
}

app.UseRouting();

if (isAuthenticationConfigured)
{
    app.UseAuthentication();
}

if (isAuthorizationConfigured)
{
    app.UseAuthorization();
}

// user middleware/endpoints
app.CustomMiddleware(...);
app.MapGet("/", () => "hello world");
// end user middleware/endpoints

app.UseEndpoints(e => {});
```

> In some cases, this default middleware configuration will not be entirely correct for the application. In which case specific sections can be replaced. For example, CORS should run before auth, in order to do that call `UseAuthentication` and `UseAuthorization` manually after `UseCors`.

Reviewer: ```suggestion
In some cases, the default middleware configuration is not correct for the app and requires modification. For example, `UseCors` should be called before `UseAuthentication` and`UseAuthorization`.
```

```csharp
app.UseCors();
app.UseAuthentication();
app.UseAuthorization();
```

> If middleware should be run before route matching occurs, `UseRouting` can be called manually and the middleware should be placed before the call to `UseRouting`. `UseEndpoints` isn't required in this case as it will still be automatically added as described above.

Reviewer: ```suggestion
If middleware should be run before route matching occurs, `UseRouting` should be called and the middleware should be placed before the call to `UseRouting`. `UseEndpoints` isn't required in this case as it is automatically added as described above.
```

```csharp
app.Use((context, next) =>
{
    return next(context);
});

app.UseRouting();

// other middleware and endpoints
```

> Another case is adding a terminal middleware (middleware that runs if no endpoint handles the request). To do this, the middleware must be added after `UseEndpoints` which means the application needs to manually call `UseRouting` and `UseEndpoints` so that the terminal middleware can be placed at the correct location.

Reviewer: ```suggestion
When adding a terminal middleware (middleware that runs if no endpoint handles the request):

* The middleware must be added after `UseEndpoints`.
* The app needs to call `UseRouting` and `UseEndpoints` so that the terminal middleware can be placed at the correct location.
```
